### PR TITLE
Updating windows to ffmpeg 4.1 and link all handled lib versions

### PR DESCRIPTION
**Pull request by ianshmean**
Currently build completes without error, but `] test VideoIO` fails because ffmpeg isn't present in the VideoIO deps folder

### CODE REVIEW

**Comment by ianshmean** [comment]
@kmsquire I've pulled your appveyor changes in and making some progress, but `Base.close_pipe_sync(cmd_out)` below is throwing an error as the [input type has changed in base](https://github.com/JuliaLang/julia/blob/0680f19907051ba95046397367b3de2bb1afe6e0/base/stream.jl#L691-L701)
https://github.com/JuliaIO/VideoIO.jl/blob/e5ee07f50b313265be7b7adbbea88fff2ba1f334/src/util.jl#L31-L50

Perhaps the function needs to be rewritten more like the underlying `link_pipe!()` structure?  
https://github.com/JuliaLang/julia/blob/0680f19907051ba95046397367b3de2bb1afe6e0/base/stream.jl#L660-L678

**Comment by kmsquire** [comment]
Yes, that definitely needs to be updated. At the time I wrote that, base only had rudimentary support for reading the output of external commands, and I had to jump through hoops to get it to handle reading both stdout and stderr from the same process. That has surely improved, so this function might not be necessary at all.

File: src/avio.jl [hunk 1]
```diff
@@ -631,8 +631,11 @@ if have_avdevice()
         camera_devices = String[]
 
         read_vid_devs = false
-        out, err = readall_stdout_stderr(`$ffmpeg -list_devices true -f $idev -i $idev_name`)
-        buf = length(out) > 0 ? out : err
+        #out, err = readall_stdout_stderr(`$ffmpeg -list_devices true -f $idev -i $idev_name`)
+        #buf = length(out) > 0 ? out : err
+        
+        out = open(`$ffmpeg -list_devices true -f $idev -i $idev_name`, other=devnull)
```
Comment (kmsquire) [inline src/avio.jl]: `other` is a positional argument, not a keyword argument.

**Comment by ianshmean** [comment]
Ok, I'm trying replacing it. Building from scratch takes a very long time on Windows. Especially GTK. 
By the way, BinDeps has some <v1.0 bugs that've had to fix https://github.com/JuliaPackaging/BinDeps.jl/pull/395

File: src/avio.jl (hunk 1 above)
Comment (ianshmean) [reply src/avio.jl]: My mistake, now corrected

Comment (kmsquire) [reply src/avio.jl]: After seeing the latest error, I remembered that `avio.jl` redefines `open` without importing `Base.open` (on purpose).  You'll need to call this with `Base.open()`

File: src/avio.jl [hunk 2]
```diff
@@ -631,8 +631,11 @@ if have_avdevice()
         camera_devices = String[]
 
         read_vid_devs = false
-        out, err = readall_stdout_stderr(`$ffmpeg -list_devices true -f $idev -i $idev_name`)
-        buf = length(out) > 0 ? out : err
+        #out, err = readall_stdout_stderr(`$ffmpeg -list_devices true -f $idev -i $idev_name`)
+        #buf = length(out) > 0 ? out : err
+        
+        out = open(`$ffmpeg -list_devices true -f $idev -i $idev_name`, devnull)
+        buf = length(out) > 0 ? out : Base.stderr
```
Comment (kmsquire) [inline src/avio.jl]: How about trying something like this:

```julia
stdout_buf = IOBuffer()
stderr_buf = IOBuffer()
run(pipeline(ignorestatus(`$ffmpeg -list_devices true -f $idev -i $idev_name`), stdout=stdout_buf, stderr=stderr_buf))
buf = (position(stdout_buf) == 0) ? stdout_buf : stderr_buf
seekstart(buf)

for line in eachline(buf)
   ...
```

There might be a more elegant way to do this, but this seems like it should work.

Comment (ianshmean) [reply src/avio.jl]: Ok. I can't test at my end right now, but I've made the change and will see how appveyor does 👍

File: src/avio.jl [hunk 3]
```diff
@@ -631,9 +631,13 @@ if have_avdevice()
         camera_devices = String[]
 
         read_vid_devs = false
-        out, err = readall_stdout_stderr(`$ffmpeg -list_devices true -f $idev -i $idev_name`)
-        buf = length(out) > 0 ? out : err
-        for line in eachline(IOBuffer(buf))
+        stdout_buf = IOBuffer()
+        stderr_buf = IOBuffer()
+        run(pipeline(ignorestatus(`$ffmpeg -list_devices true -f $idev -i $idev_name`), stdout=stdout_buf, stderr=stderr_buf))
```
Comment (ianshmean) [inline src/avio.jl]: erroring here with `no method matching rawhandle(::Base.GenericIOBuffer{Array{UInt8,1}})`

**Comment by ianshmean** [comment]
Perhaps the solution lies in these changes?  https://github.com/JuliaLang/julia/pull/30278/files

File: src/avio.jl (hunk 3 above)
Comment (ianshmean) [reply src/avio.jl]: This works on windows on 1.1.0 (but not 1.0.3).
I hadn't appreciated it was a new change

**Comment by ianshmean** [comment]
I got `BinDeps.debug("VideoIO")` working with https://github.com/JuliaPackaging/BinDeps.jl/pull/395
```
julia> BinDeps.debug("VideoIO")
[ Info: Reading build script...
┌ Warning: `Pkg.dir(pkgname, paths...)` is deprecated; instead, do `import VideoIO; joinpath(dirname(pathof(VideoIO)), "..", paths...)`.
└ @ Pkg.API C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.1\Pkg\src\API.jl:395
The package declares 1 dependencies.
 - Library Group "libav" (satisfied by BinDeps.SystemPaths)
     - Library "libavcodec"
        - Satisfied by:
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avcodec-57.DLL
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avcodec-57.dll
        - Providers:
          - Binaries
          - AptGet package libavcodec55-ffmpeg (can't provide)
          - AptGet package libavcodec57 (can't provide)
          - AptGet package libavcodec-extra-54 (can't provide)
          - AptGet package libavcodec54 (can't provide)
          - AptGet package libavcodec56 (can't provide)
          - AptGet package libavcodec-extra-53 (can't provide)
          - AptGet package libavcodec-extra-56 (can't provide)
          - AptGet package libavcodec-extra-58 (can't provide)
          - AptGet package libavcodec58 (can't provide)
          - AptGet package libavcodec53 (can't provide)
          - AptGet package libavcodec55 (can't provide)
          - AptGet package libavcodec-extra-57 (can't provide)
          - AptGet package libavcodec-extra-55 (can't provide)
          - AptGet package libavcodec-ffmpeg56 (can't provide)
          - Yum package ffmpeg (can't provide)
          - Pacman package ffmpeg (can't provide)
          - Autotools Build (can't provide)
     - Library "libavformat"
        - Satisfied by:
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avformat-57.DLL
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avformat-57.dll
        - Providers:
          - Binaries
          - AptGet package libavformat56 (can't provide)
          - AptGet package libavformat53 (can't provide)
          - AptGet package libavformat55 (can't provide)
          - AptGet package libavformat54 (can't provide)
          - AptGet package libavformat57 (can't provide)
          - AptGet package libavformat55-ffmpeg (can't provide)
          - AptGet package libavformat58 (can't provide)
          - AptGet package libavformat-ffmpeg56 (can't provide)
          - Yum package ffmpeg (can't provide)
          - Pacman package ffmpeg (can't provide)
          - Autotools Build (can't provide)
     - Library "libavutil"
        - Satisfied by:
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avutil-55.DLL
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avutil-55.dll
        - Providers:
          - Binaries
          - AptGet package libavutil55 (can't provide)
          - AptGet package libavutil-ffmpeg54 (can't provide)
          - AptGet package libavutil52 (can't provide)
          - AptGet package libavutil53 (can't provide)
          - AptGet package libavutil51 (can't provide)
          - AptGet package libavutil56 (can't provide)
          - AptGet package libavutil52-ffmpeg (can't provide)
          - AptGet package libavutil54 (can't provide)
          - Yum package ffmpeg (can't provide)
          - Pacman package ffmpeg (can't provide)
          - Autotools Build (can't provide)
     - Library "libswscale"
        - Satisfied by:
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\swscale-4.DLL
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\swscale-4.dll
        - Providers:
          - Binaries
          - AptGet package libswscale2 (can't provide)
          - AptGet package libswscale3 (can't provide)
          - AptGet package libswscale2-ffmpeg (can't provide)
          - AptGet package libswscale4 (can't provide)
          - AptGet package libswscale-ffmpeg3 (can't provide)
          - AptGet package libswscale5 (can't provide)
          - Yum package ffmpeg (can't provide)
          - Pacman package ffmpeg (can't provide)
          - Autotools Build (can't provide)
     - Library "libavfilter"
        - Satisfied by:
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avfilter-6.DLL
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avfilter-6.dll
        - Providers:
          - Binaries
          - AptGet package libavfilter7 (can't provide)
          - AptGet package libavfilter-ffmpeg5 (can't provide)
          - AptGet package libavfilter6 (can't provide)
          - AptGet package libavfilter5 (can't provide)
          - AptGet package libavfilter4 (can't provide)
          - AptGet package libavfilter2 (can't provide)
          - AptGet package libavfilter4-ffmpeg (can't provide)
          - AptGet package libavfilter3 (can't provide)
          - Yum package ffmpeg (can't provide)
          - Pacman package ffmpeg (can't provide)
          - Autotools Build (can't provide)
     - Library "libavdevice"
        - Satisfied by:
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avdevice-57.DLL
          - System Paths at C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\avdevice-57.dll
        - Providers:
          - Binaries
          - AptGet package libavdevice54 (can't provide)
          - AptGet package libavdevice57 (can't provide)
          - AptGet package libavdevice53 (can't provide)
          - AptGet package libavdevice58 (can't provide)
          - AptGet package libavdevice55 (can't provide)
          - AptGet package libavdevice56 (can't provide)
          - AptGet package libavdevice55-ffmpeg (can't provide)
          - AptGet package libavdevice-ffmpeg56 (can't provide)
          - Yum package ffmpeg (can't provide)
          - Pacman package ffmpeg (can't provide)
          - Autotools Build (can't provide)
```
Given the libs are satisfied, is it then invalid to be looking in `deps\`?
```
(v1.1) pkg> test VideoIO
...
ERROR: LoadError: LoadError: LoadError: LoadError: IOError: could not spawn `'C:\Users\Ian\Documents\GitHub\VideoIO.jl\src\..\deps\ffmpeg-4.1-win64-shared\bin\ffmpeg.exe' -list_devices true -f dshow -i dummy`: no such file or directory (ENOENT)
Stacktrace:
...
```

**Comment by ianshmean** [comment]
Seem to be making some progress on appveyor.. If we're supporting <1.1 the IOBuffer() method needs work

**Comment by kmsquire** [comment]
#140 should (hopefully!) fix the 32-bit error on Windows.  We don't test 32-bit builds on other platforms, so it's nice to be able to do both on Windows.

**Comment by ianshmean** [comment]
@kmsquire I actually relaxed that too in the last commit. Perhaps a bit too far, because I abstracted the floats too. If it passes I can revert to the form in #140 but given how long it takes to run appveyor, I just want to see it working for now... 30-45 mins per test... mostly on dep builds!

**Comment by kmsquire** [comment]
> I actually relaxed that too in the last commit.

Sorry, missed that.  I'll close #140, then.

> Given the libs are satisfied, is it then invalid to be looking in `deps\`?

The answer would seem to be "no". ;-)

https://github.com/JuliaIO/VideoIO.jl/blob/aea923dfcf066e72ce058aecb5484babdec4833a/src/avio.jl#L667 should probably use `INSTALL_ROOT`, as is done at https://github.com/JuliaIO/VideoIO.jl/blob/aea923dfcf066e72ce058aecb5484babdec4833a/src/avio.jl#L683, assuming the location of the binary is deterministically related to the location of the libraries.

**Comment by kmsquire** [comment]
@ianshmean I just sent you an invite to join the `JuliaIO/video` team (which I just created).  If you join, you should be able to cancel Appveyor jobs (assuming I set it up properly).

(I separately sent an invite from Appveyor, but I think the `JuliaIO/video` team is the better option, so I canceled the other one.)

**Comment by ianshmean** [comment]
Great! I'm in. Thanks
Unfortunately on my local windows build, the ffmpeg binary is nowhere to be found around the `C:\Program Files\Point Grey Research\Spinnaker\bin64\vs2015\` location of the libs. So perhaps build should be modified to install the binary in `deps\`, even if all the libs are satisfied?

**Comment by kmsquire** [comment]
> So perhaps build should be modified to install the binary in `deps\`, even if all the libs are satisfied?

That generally sounds fine.  Does the windows binary statically link the libraries, or does it depend on the `dll`s being installed?  If the latter, maybe it's best to always install everything on Windows (although I have no idea how to do that with `BinDeps`.)

**Comment by kmsquire** [comment]
Also, there may be a more nuanced way of handling this, but let's try caching some of the build dependencies.  Can you try addin this to `appveyor.yml`:

```
cache:
  - C:\Users\appveyor\.julia\packages\Homebrew
  - C:\Users\appveyor\.julia\packages\LibCURL
  - C:\Users\appveyor\.julia\packages\WinRPM
  - C:\Users\appveyor\.julia\packages\Cairo
  - C:\Users\appveyor\.julia\packages\Conda
  - C:\Users\appveyor\.julia\packages\FFTW
  - C:\Users\appveyor\.julia\packages\SpecialFunctions
  - C:\Users\appveyor\.julia\packages\Gtk
  - C:\Users\appveyor\.julia\packages\ImageMagick
```

I've never done this, so I don't know if it will work, but hopefully caching these will speed up builds.  We only get 1GB of space for all builds (files are compressed), though, and I'm not sure how big these directories are...  I also don't know where things like `dll`s are put on Windows--are they somewhere in the package directory?  If not, then those locations will probably need to be added above.

**Comment by ianshmean** [comment]
I've added the caching. Good idea.
btw, the first 1.1 build on windows just passed on appveyor. progress!

**Comment by ianshmean** [comment]
Opened a discussion on the stdout stderr handling for 1.0 
https://discourse.julialang.org/t/capture-stdout-and-stderr-in-1-1/20712

**Comment by ianshmean** [comment]
I think we're there, but the appveyor cache hasn't worked though so I'll check back tomorrow. Given the built packages are in randomly named subdirs of the provided dirs, I'm guessing the fresh package registry doesn't know they're there.

**Comment by kmsquire** [comment]
> I think we're there, but enabling the cache in appveyor hasn't worked as expected, so I'll check back tomorrow. Given the built packages are in randomly named subdirs of the provided dirs, I'm guessing the fresh package registry doesn't know they're there?

No, I just figured out that I had to enable saving the cache in the project settings.  I just did that, so hopefully we'll be caching these packages going forward.
(The random characters after the name are deterministic... I'm not sure why they're there, though.)

File: src/util.jl [hunk 4]
```diff
@@ -28,28 +28,28 @@ end
...
+#    Base.start_reading(out)
+#    Base.start_reading(err)
+#
+#    return (out, err, r)
+#end
+#
+#function readall_stdout_stderr(cmd::Cmd)
+#    (out, err, proc) = open_stdout_stderr(cmd)
+#    return (read(out, String), read(err, String))
+#end
```
Comment (kmsquire) [inline src/util.jl]: Let's just delete these lines.  (No reason to run CI again.)

Comment (ianshmean) [reply src/util.jl]: Indeed. I've removed them, but actually don't see options to cancel/edit CI jobs on appveyor, even though I'm logged in with access granted to the VideoIO org. Anyhow, would you mind cancelling the new jobs for the moment? thanks

**Comment by ianshmean** [comment]
CI tests pass, but there is the outstanding issue on my local windows build, where the libs are already present but the binary isn't. Perhaps that's a separate issue that can wait for #120 
Shall we merge @kmsquire ?

**Comment by kmsquire** [comment]
I squashed and cleaned up the commit history a little.  (I probably should have asked if you wanted to do this, or at least warn you--sorry!)

No functional changes (or at least there shouldn't be).

I've updated the Appveyor access for @JuliaIO/video to Administrator, so hopefully you should be able to cancel Appveyor builds now.  I'll be curious to see if the caching works this time.

I think this can be merged as soon as at least one of the Appveyor jobs completes.

**Comment by ianshmean** [comment]
Thanks for the tidyup. Shall we release a version, given the previous one was nonfunctional on windows? (Just checking I'm not missing something)

**Comment by kmsquire** [comment]
Please do!

**Comment by ianshmean** [comment]
Done, thanks

**Pull request merged**